### ISSUE TITLE
TiKV OOM under high write pressure

### ISSUE THREAD

**Issue by cosven** [tikv/tikv#10445]
### What version of TiKV are you using?

https://github.com/tikv/tikv/commit/2ab5b40b2bfdceaeaf61806ffe5e24b57d5c3ebb
```
[root@7b6cbaa65e59 log]# ../bin/tikv-server -V
TiKV
Release Version:   5.1.0
Edition:           Community
Git Commit Hash:   2ab5b40b2bfdceaeaf61806ffe5e24b57d5c3ebb
Git Commit Branch: heads/refs/tags/v5.1.0
UTC Build Time:    2021-06-22 12:17:22
Rust Version:      rustc 1.53.0-nightly (16bf626a3 2021-04-14)
Enable Features:   jemalloc mem-profiling portable sse protobuf-codec test-engines-rocksdb cloud-aws cloud-gcp
Profile:           dist_release
```

### Steps to reproduce

1. Run write-stress testcase.
  `/tmp/hhhhh/bin/write-stress -tidb-server xxx -case-name append -tables 128 -pad-length 4000000 -run-time 3h -concurrency 100`
2. The TiKV may run into OOM.

### What did you expect?

No OOM.

### What did happened?

OOM.

Note: the metrics shows that the entry cache does not take up too much memory.
![image](https://user-images.githubusercontent.com/4962134/123115273-8e147e00-d472-11eb-83bb-e87baa9b9dba.png)
![image](https://user-images.githubusercontent.com/4962134/123115450-b308f100-d472-11eb-95af-1912980b24b9.png)

**Comment by cosven** [comment]
/severity major
/type bug
/assign @hicqu 

Related issue: https://github.com/tikv/tikv/issues/10444

**Comment by hicqu** [comment]
Fixed by tikv/tikv#10334.
